src/conversion/viewconversiondispatcher.js

```diff
@@ -217,19 +250,77 @@ export default class ViewConversionDispatcher {
 	 * @private
 	 * @see module:engine/conversion/viewconversiondispatcher~ViewConversionApi#convertChildren
 	 */
-	_convertChildren( input, consumable, additionalData ) {
-		const writer = this.conversionApi.writer;
-		const documentFragment = writer.createDocumentFragment();
+	_convertChildren( input, consumable, position ) {
+		const output = new ModelRange( position );
 
 		for ( const viewChild of Array.from( input.getChildren() ) ) {
-			const modelChild = this._convertItem( viewChild, consumable, additionalData );
+			const range = this._convertItem( viewChild, consumable, output.end );
 
-			if ( modelChild instanceof ModelNode || modelChild instanceof ModelDocumentFragment ) {
-				writer.append( modelChild, documentFragment );
+			if ( range instanceof ModelRange ) {
+				output.end = range.end;
 			}
 		}
 
-		return documentFragment;
+		return output;
+	}
+
+	/**
+	 * @private
+	 * @see module:engine/conversion/viewconversiondispatcher~ViewConversionApi#splitToAllowedParent
+	 */
+	_splitToAllowedParent( element, position ) {
+		// Try to find allowed parent.
+		const allowedParent = this.conversionApi.schema.findAllowedParent( element, position, this._contextPosition.parent );
+
+		// When there is no parent that allows to insert element then return `null`.
+		if ( !allowedParent ) {
+			return null;
+		}
+
+		// When current position parent allows to insert element then return this position.
+		if ( allowedParent === position.parent ) {
+			return { position };
+		}
+
+		// Split element to allowed parent.
+		const data = this.conversionApi.writer.split( position, allowedParent );
+
+		// Remember all elements that are created as a result of split.
+		// This is important because at the end of conversion we want to remove all empty split elements.
+		for ( const pos of data.range.getPositions() ) {
+			if ( !pos.isEqual( data.position ) ) {
+				this._splitElements.add( pos.parent );
```

Comment:
Maybe `_splitElements` -> `_removeIfEmpty`. It does not matter anymore that these elements are results of the split. What does matter is that they should be removed if empty. Also, in the future, we may have more elements like this.